torchaudio/functional/functional.py

```diff
@@ -971,3 +975,34 @@ def spectral_centroid(
                            device=specgram.device).reshape((-1, 1))
     freq_dim = -2
     return (freqs * specgram).sum(dim=freq_dim) / specgram.sum(dim=freq_dim)
+
+
+def apply_codec(waveform, sample_rate, compression, format, channels_first=True) -> Tensor:
+    r"""
+    Applies codecs as a form of augmentation
+
+    Args:
+        waveform (Tensor): Audio data. Must be 2 dimensional. See also ```channels_first```
+        sample_rate (int): Sample rate of the audio waveform
+        format (str): file format
+        channels_first (bool):
+            When True, the returned Tensor has dimension ``[channel, time]``.
+            Otherwise, the returned Tensor's dimension is ``[time, channel]``.
+        compression (float):
+            Used for formats other than WAV. This corresponds to ``-C`` option of ``sox`` command.
+                * | ``MP3``: Either bitrate (in ``kbps``) with quality factor, such as ``128.2``, or
+                  | VBR encoding with quality factor such as ``-4.2``. Default: ``-4.5``.
+                * | ``FLAC``: compression level. Whole number from ``0`` to ``8``.
+                  | ``8`` is default and highest compression.
+                * | ``OGG/VORBIS``: number from ``-1`` to ``10``; ``-1`` is the highest compression
+                  | and lowest quality. Default: ``3``.
+            See the detail at http://sox.sourceforge.net/soxformat.html.
+
+    Returns:
+        Tensor
+    """
+    bytes = io.BytesIO()
+    torchaudio.save(bytes, waveform, sample_rate, channels_first, compression=compression, format=format)
+    bytes.seek(0)
+    waveform, _ = torchaudio.load(bytes, channels_first=channels_first)
```

Comment:
I realized that there are certain formats that has limitation on sample rate.

Can you replace `load` with `sox_effects.apply_effects_file`, and provide `[["rate", f"{sample_rate}"]]` so that they are re-sampled to original sample rate if necessary?

 ```python
augmented, _ = torchaudio.sox_effect.apply_effects_file(
    bytes, effects=[["rate", f"{sample_rate}"]], channels_first=channels_first, format=format)
return augmented
```